Regression in v2.2.5 - pion/webrtc simulcast tests fail

I was trying to bump versions of all pion modules uses by v3 of `pion/webrtc`, and found that simulcast tests in pion/webrtc test suite fails with v2.2.5. They work fine with v2.2.4. Here is part of log for failing tests:

```
  ❌ TestPeerConnection_Simulcast_Probe (0s)
  ❌ TestPeerConnection_Simulcast_Probe/Break_NonSimulcast (0s)
goroutine profile: total 42
4 @ 0x8085988 0x80968a8 0x837c0e1 0x84b6f4c 0x83f4d20 0x80b7c81
#	0x837c0e0	github.com/pion/transport/v2/packetio.(*Buffer).Read+0x140	/go/pkg/mod/github.com/pion/transport/v2@v2.2.5/packetio/buffer.go:265
#	0x84b6f4b	github.com/pion/webrtc/v3/internal/mux.(*Endpoint).Read+0x3b	/go/src/github.com/pion/webrtc/internal/mux/endpoint.go:40
#	0x83f4d1f	github.com/pion/srtp/v2.(*session).start.func1+0xbf		/go/pkg/mod/github.com/pion/srtp/v2@v2.0.20/session.go:144

4 @ 0x8085988 0x80b5324 0x80b5300 0x80be490 0x83d660b 0x83dcdd0 0x84cab7e 0x80b7c81
#	0x80b52ff	sync.runtime_notifyListWait+0x14f				/usr/local/go/src/runtime/sema.go:517
#	0x80be48f	sync.(*Cond).Wait+0x9f						/usr/local/go/src/sync/cond.go:70
#	0x83d660a	github.com/pion/sctp.(*Stream).ReadSCTP+0xca			/go/pkg/mod/github.com/pion/sctp@v1.8.19/stream.go:146
#	0x83dcdcf	github.com/pion/datachannel.(*DataChannel).ReadDataChannel+0x3f	/go/pkg/mod/github.com/pion/datachannel@v1.5.8/datachannel.go:198
#	0x84cab7d	github.com/pion/webrtc/v3.(*DataChannel).readLoop+0xcd		/go/src/github.com/pion/webrtc/datachannel.go:361

2 @ 0x8085988 0x804dee2 0x804dadc 0x83dc82b 0x83dc80a 0x84f8246 0x80b7c81
#	0x83dc82a	github.com/pion/sctp.(*Association).AcceptStream+0x3a			/go/pkg/mod/github.com/pion/sctp@v1.8.19/association.go:1513
#	0x83dc809	github.com/pion/datachannel.Accept+0x19					/go/pkg/mod/github.com/pion/datachannel@v1.5.8/datachannel.go:131
#	0x84f8245	github.com/pion/webrtc/v3.(*SCTPTransport).acceptDataChannels+0x1e5	/go/src/github.com/pion/webrtc/sctptransport.go:179

2 @ 0x8085988 0x804dee2 0x804dadc 0x83f55d0 0x84e8579 0x80b7c81
#	0x83f55cf	github.com/pion/srtp/v2.(*SessionSRTCP).AcceptStream+0x3f			/go/pkg/mod/github.com/pion/srtp/v2@v2.0.20/session_srtcp.go:96
#	0x84e8578	github.com/pion/webrtc/v3.(*PeerConnection).undeclaredRTCPMediaProcessor+0xb8	/go/src/github.com/pion/webrtc/peerconnection.go:1690

2 @ 0x8085988 0x804dee2 0x804dadc 0x83f6550 0x84e7f8a 0x80b7c81
#	0x83f654f	github.com/pion/srtp/v2.(*SessionSRTP).AcceptStream+0x3f			/go/pkg/mod/github.com/pion/srtp/v2@v2.0.20/session_srtp.go:98
#	0x84e7f89	github.com/pion/webrtc/v3.(*PeerConnection).undeclaredRTPMediaProcessor+0x79	/go/src/github.com/pion/webrtc/peerconnection.go:1646

2 @ 0x8085988 0x807daaa 0x80b39f2 0x80ea127 0x80eb75f 0x80eb744 0x82c6887 0x82e51cd 0x82e344f 0x82e35f3 0x849cb84 0x80b7c81
#	0x80b39f1	internal/poll.runtime_pollWait+0x51			/usr/local/go/src/runtime/netpoll.go:305
#	0x80ea126	internal/poll.(*pollDesc).wait+0x36			/usr/local/go/src/internal/poll/fd_poll_runtime.go:84
#	0x80eb75e	internal/poll.(*pollDesc).waitRead+0x1be		/usr/local/go/src/internal/poll/fd_poll_runtime.go:89
#	0x80eb743	internal/poll.(*FD).ReadFromInet4+0x1a3			/usr/local/go/src/internal/poll/fd_unix.go:250
#	0x82c6886	net.(*netFD).readFromInet4+0x46				/usr/local/go/src/net/fd_posix.go:66
#	0x82e51cc	net.(*UDPConn).readFrom+0x1cc				/usr/local/go/src/net/udpsock_posix.go:52
#	0x82e344e	net.(*UDPConn).readFromUDP+0x4e				/usr/local/go/src/net/udpsock.go:149
#	0x82e35f2	net.(*UDPConn).ReadFrom+0x52				/usr/local/go/src/net/udpsock.go:158
#	0x849cb83	github.com/pion/ice/v2.(*candidateBase).recvLoop+0x153	/go/pkg/mod/github.com/pion/ice/v2@v2.3.29/candidate_base.go:227

2 @ 0x8085988 0x807daaa 0x80b39f2 0x80ea127 0x80eed91 0x80eed72 0x82dd08f 0x836d3a1 0x8377ca2 0x8377c88 0x8412ed4 0x80b7c81
#	0x80b39f1	internal/poll.runtime_pollWait+0x51			/usr/local/go/src/runtime/netpoll.go:305
#	0x80ea126	internal/poll.(*pollDesc).wait+0x36			/usr/local/go/src/internal/poll/fd_poll_runtime.go:84
#	0x80eed90	internal/poll.(*pollDesc).waitRead+0x140		/usr/local/go/src/internal/poll/fd_poll_runtime.go:89
#	0x80eed71	internal/poll.(*FD).RawRead+0x121			/usr/local/go/src/internal/poll/fd_unix.go:766
#	0x82dd08e	net.(*rawConn).Read+0x3e				/usr/local/go/src/net/rawconn.go:43
#	0x836d3a0	golang.org/x/net/internal/socket.(*Conn).recvMsg+0x180	/go/pkg/mod/golang.org/x/net@v0.22.0/internal/socket/rawconn_msg.go:27
#	0x8377ca1	golang.org/x/net/internal/socket.(*Conn).RecvMsg+0x451	/go/pkg/mod/golang.org/x/net@v0.22.0/internal/socket/socket.go:247
#	0x8377c87	golang.org/x/net/ipv4.(*payloadHandler).ReadFrom+0x437	/go/pkg/mod/golang.org/x/net@v0.22.0/ipv4/payload_cmsg.go:31
#	0x8412ed3	github.com/pion/mdns.(*Conn).start+0xc3			/go/pkg/mod/github.com/pion/mdns@v0.0.12/conn.go:407

2 @ 0x8085988 0x80968a8 0x8140dd6 0x80b7c81
#	0x8140dd5	context.propagateCancel.func1+0xb5	/usr/local/go/src/context/context.go:279

2 @ 0x8085988 0x80968a8 0x8366a69 0x80b7c81
#	0x8366a68	github.com/pion/transport/v2/connctx.(*connCtx).ReadContext.func1+0xd8	/go/pkg/mod/github.com/pion/transport/v2@v2.2.5/connctx/connctx.go:83

2 @ 0x8085988 0x80968a8 0x837c0e1 0x84b2062 0x84b780a 0x80b7c81
#	0x837c0e0	github.com/pion/transport/v2/packetio.(*Buffer).Read+0x140	/go/pkg/mod/github.com/pion/transport/v2@v2.2.5/packetio/buffer.go:265
#	0x84b2061	github.com/pion/ice/v2.(*Conn).Read+0xa1			/go/pkg/mod/github.com/pion/ice/v2@v2.3.29/transport.go:76
#	0x84b7809	github.com/pion/webrtc/v3/internal/mux.(*Mux).readLoop+0xa9	/go/src/github.com/pion/webrtc/internal/mux/mux.go:110

2 @ 0x8085988 0x80968a8 0x837c0e1 0x84b6f4c 0x8366887 0x839d669 0x839fdba 0x80b7c81
#	0x837c0e0	github.com/pion/transport/v2/packetio.(*Buffer).Read+0x140		/go/pkg/mod/github.com/pion/transport/v2@v2.2.5/packetio/buffer.go:265
#	0x84b6f4b	github.com/pion/webrtc/v3/internal/mux.(*Endpoint).Read+0x3b		/go/src/github.com/pion/webrtc/internal/mux/endpoint.go:40
#	0x8366886	github.com/pion/transport/v2/connctx.(*connCtx).ReadContext+0x1e6	/go/pkg/mod/github.com/pion/transport/v2@v2.2.5/connctx/connctx.go:98
#	0x839d668	github.com/pion/dtls/v2.(*Conn).readAndBuffer+0x118			/go/pkg/mod/github.com/pion/dtls/v2@v2.2.12/conn.go:604
#	0x839fdb9	github.com/pion/dtls/v2.(*Conn).handshake.func3+0xf9			/go/pkg/mod/github.com/pion/dtls/v2@v2.2.12/conn.go:891

2 @ 0x8085988 0x80968a8 0x839b76b 0x83b848f 0x80b7c81
#	0x839b76a	github.com/pion/dtls/v2.(*Conn).Read+0x14a		/go/pkg/mod/github.com/pion/dtls/v2@v2.2.12/conn.go:325
#	0x83b848e	github.com/pion/sctp.(*Association).readLoop+0x15e	/go/pkg/mod/github.com/pion/sctp@v1.8.19/association.go:594

2 @ 0x8085988 0x80968a8 0x83b2312 0x83b13b7 0x83a03e8 0x80b7c81
#	0x83b2311	github.com/pion/dtls/v2.(*handshakeFSM).finish+0x221	/go/pkg/mod/github.com/pion/dtls/v2@v2.2.12/handshaker.go:322
#	0x83b13b6	github.com/pion/dtls/v2.(*handshakeFSM).Run+0x3b6	/go/pkg/mod/github.com/pion/dtls/v2@v2.2.12/handshaker.go:189
#	0x83a03e7	github.com/pion/dtls/v2.(*Conn).handshake.func2+0x97	/go/pkg/mod/github.com/pion/dtls/v2@v2.2.12/conn.go:872

2 @ 0x8085988 0x80968a8 0x83b8f37 0x80b7c81
#	0x83b8f36	github.com/pion/sctp.(*Association).writeLoop+0x1f6	/go/pkg/mod/github.com/pion/sctp@v1.8.19/association.go:644

2 @ 0x8085988 0x80968a8 0x84927d4 0x80b7c81
#	0x84927d3	github.com/pion/ice/v2.(*Agent).taskLoop+0x143	/go/pkg/mod/github.com/pion/ice/v2@v2.3.29/agent.go:239

2 @ 0x8085988 0x80968a8 0x84942d3 0x80b7c81
#	0x84942d2	github.com/pion/ice/v2.(*Agent).connectivityChecks+0x322	/go/pkg/mod/github.com/pion/ice/v2@v2.3.29/agent.go:481

1 @ 0x8079a19 0x80b35a9 0x81d8bf2 0x81d8a1a 0x81d5465 0x84c6437 0x80b7c81
#	0x80b35a8	runtime/pprof.runtime_goroutineProfileWithLabels+0x48	/usr/local/go/src/runtime/mprof.go:846
#	0x81d8bf1	runtime/pprof.writeRuntimeProfile+0xc1			/usr/local/go/src/runtime/pprof/pprof.go:723
#	0x81d8a19	runtime/pprof.writeGoroutine+0x79			/usr/local/go/src/runtime/pprof/pprof.go:683
#	0x81d5464	runtime/pprof.(*Profile).WriteTo+0xe4			/usr/local/go/src/runtime/pprof/pprof.go:330
#	0x84c6436	github.com/pion/transport/v2/test.TimeOut.func1+0x56	/go/pkg/mod/github.com/pion/transport/v2@v2.2.5/test/util.go:24

1 @ 0x8085988 0x804dee2 0x804dabc 0x8151913 0x815376e 0x8150b93 0x8153656 0x81521f0 0x85861b5 0x808558d 0x80b7c81
#	0x8151912	testing.(*T).Run+0x392		/usr/local/go/src/testing/testing.go:1494
#	0x815376d	testing.runTests.func1+0x5d	/usr/local/go/src/testing/testing.go:1846
#	0x8150b92	testing.tRunner+0x112		/usr/local/go/src/testing/testing.go:1446
#	0x8153655	testing.runTests+0x3a5		/usr/local/go/src/testing/testing.go:1844
#	0x81521ef	testing.(*M).Run+0x67f		/usr/local/go/src/testing/testing.go:1726
#	0x85861b4	main.main+0x184			_testmain.go:541
#	0x808558c	runtime.main+0x24c		/usr/local/go/src/runtime/proc.go:250

1 @ 0x8085988 0x804dee2 0x804dabc 0x8151913 0x8543e27 0x8150b93 0x80b7c81
#	0x8151912	testing.(*T).Run+0x392							/usr/local/go/src/testing/testing.go:1494
#	0x8543e26	github.com/pion/webrtc/v3.TestPeerConnection_Simulcast_Probe+0xd6	/go/src/github.com/pion/webrtc/peerconnection_media_test.go:1050
#	0x8150b92	testing.tRunner+0x112							/usr/local/go/src/testing/testing.go:1446

1 @ 0x8085988 0x80968a8 0x837c0e1 0x84cf1ef 0x84cf1cc 0x8212b0c 0x84e76bf 0x84e83b7 0x80b7c81
#	0x837c0e0	github.com/pion/transport/v2/packetio.(*Buffer).Read+0x140				/go/pkg/mod/github.com/pion/transport/v2@v2.2.5/packetio/buffer.go:265
#	0x84cf1ee	github.com/pion/srtp/v2.(*ReadStreamSRTP).Read+0x3e					/go/pkg/mod/github.com/pion/srtp/v2@v2.0.20/stream_srtp.go:79
#	0x84cf1cb	github.com/pion/webrtc/v3.(*DTLSTransport).streamsForSSRC.func1+0x1b			/go/src/github.com/pion/webrtc/dtlstransport.go:494
#	0x8212b0b	github.com/pion/interceptor.RTPReaderFunc.Read+0x3b					/go/pkg/mod/github.com/pion/interceptor@v0.1.29/interceptor.go:91
#	0x84e76be	github.com/pion/webrtc/v3.(*PeerConnection).handleIncomingSSRC+0x78e			/go/src/github.com/pion/webrtc/peerconnection.go:1594
#	0x84e83b6	github.com/pion/webrtc/v3.(*PeerConnection).undeclaredRTPMediaProcessor.func1+0x46	/go/src/github.com/pion/webrtc/peerconnection.go:1668

1 @ 0x8085988 0x80968a8 0x837c0e1 0x84cf1ef 0x84cf1cc 0x8212b0c 0x84f164a 0x8509510 0x85099bd 0x84e4bfc 0x80b7c81
#	0x837c0e0	github.com/pion/transport/v2/packetio.(*Buffer).Read+0x140		/go/pkg/mod/github.com/pion/transport/v2@v2.2.5/packetio/buffer.go:265
#	0x84cf1ee	github.com/pion/srtp/v2.(*ReadStreamSRTP).Read+0x3e			/go/pkg/mod/github.com/pion/srtp/v2@v2.0.20/stream_srtp.go:79
#	0x84cf1cb	github.com/pion/webrtc/v3.(*DTLSTransport).streamsForSSRC.func1+0x1b	/go/src/github.com/pion/webrtc/dtlstransport.go:494
#	0x8212b0b	github.com/pion/interceptor.RTPReaderFunc.Read+0x3b			/go/pkg/mod/github.com/pion/interceptor@v0.1.29/interceptor.go:91
#	0x84f1649	github.com/pion/webrtc/v3.(*RTPReceiver).readRTP+0x149			/go/src/github.com/pion/webrtc/rtpreceiver.go:369
#	0x850950f	github.com/pion/webrtc/v3.(*TrackRemote).Read+0x1ef			/go/src/github.com/pion/webrtc/track_remote.go:139
#	0x85099bc	github.com/pion/webrtc/v3.(*TrackRemote).peek+0x3c			/go/src/github.com/pion/webrtc/track_remote.go:192
#	0x84e4bfb	github.com/pion/webrtc/v3.(*PeerConnection).startReceiver.func1+0x7b	/go/src/github.com/pion/webrtc/peerconnection.go:1231

1 @ 0x8085988 0x80b6136 0x8543c75 0x8543a66 0x8543a54 0x8150b93 0x80b7c81
#	0x80b6135	time.Sleep+0x145								/usr/local/go/src/runtime/time.go:195
#	0x8543c74	github.com/pion/webrtc/v3.TestPeerConnection_Simulcast_Probe.func2.2+0xc4	/go/src/github.com/pion/webrtc/peerconnection_media_test.go:1106
#	0x8543a65	github.com/pion/webrtc/v3.TestPeerConnection_Simulcast_Probe.func2.5+0x675	/go/src/github.com/pion/webrtc/peerconnection_media_test.go:1137
#	0x8543a53	github.com/pion/webrtc/v3.TestPeerConnection_Simulcast_Probe.func2+0x663	/go/src/github.com/pion/webrtc/peerconnection_media_test.go:1140
#	0x8150b92	testing.tRunner+0x112								/usr/local/go/src/testing/testing.go:1446

panic: timeout

goroutine 7004 [running]:
github.com/pion/transport/v2/test.TimeOut.func1()
	/go/pkg/mod/github.com/pion/transport/v2@v2.2.5/test/util.go:27 +0xe0
created by time.goFunc
	/usr/local/go/src/time/sleep.go:176 +0x2f
  ❌ TestPeerConnection_Simulcast_Probe/Leak (0s)
```

--- a/packetio/buffer.go
+++ b/packetio/buffer.go
@@ -173,14 +173,9 @@ func (b *Buffer) Write(packet []byte) (int, error) {
 	}
 	b.count++
 
-	waiting := b.waiting
-	b.waiting = false
-
-	if waiting {
-		select {
-		case b.notify <- struct{}{}:
-		default:
-		}
+	select {
+	case b.notify <- struct{}{}:
+	default:
 	}
 	b.mutex.Unlock()
 
@@ -244,7 +239,6 @@ func (b *Buffer) Read(packet []byte) (n int, err error) { //nolint:gocognit
 			}
 
 			b.count--
-			b.waiting = false
 			b.mutex.Unlock()
 
 			if copied < count {
@@ -257,8 +251,6 @@ func (b *Buffer) Read(packet []byte) (n int, err error) { //nolint:gocognit
 			b.mutex.Unlock()
 			return 0, io.EOF
 		}
-
-		b.waiting = true
 		b.mutex.Unlock()
 
 		select {

--- a/packetio/buffer_test.go
+++ b/packetio/buffer_test.go
@@ -8,9 +8,11 @@ import (
 	"fmt"
 	"io"
 	"net"
+	"sync/atomic"
 	"testing"
 	"time"
 
+	"github.com/pion/transport/v3/test"
 	"github.com/stretchr/testify/assert"
 )
 
@@ -611,6 +613,7 @@ func TestBufferConcurrentRead(t *testing.T) {
 
 	errCh := make(chan error, 2)
 	readIntoErr := func() {
+		packet := make([]byte, 4)
 		_, readErr := buffer.Read(packet)
 		errCh <- readErr
 	}
@@ -625,4 +628,39 @@ func TestBufferConcurrentRead(t *testing.T) {
 	assert.Equal(io.EOF, err)
 	err = <-errCh
 	assert.Equal(io.EOF, err)
+}
+
+func TestBufferConcurrentReadWrite(t *testing.T) {
+	defer test.TimeOut(time.Second * 5).Stop()
+
+	assert := assert.New(t)
+
+	buffer := NewBuffer()
+
+	numPkts := 1000
+	var numRead uint64
+	allRead := make(chan struct{})
+	readPkts := func(count int) {
+		packet := make([]byte, 4)
+		for i := 0; i < count; i++ {
+			_, readErr := buffer.Read(packet)
+			if readErr != nil {
+				return
+			}
+			if atomic.AddUint64(&numRead, 1) == uint64(numPkts) {
+				close(allRead)
+				return
+			}
+		}
+	}
+	go readPkts(numPkts)
+	go readPkts(numPkts / 100)
+
+	for i := 0; i < numPkts; i++ {
+		_, writeErr := buffer.Write([]byte{2, 3, 4})
+		assert.NoError(writeErr)
+	}
+	<-allRead
+
+	assert.NoError(buffer.Close())
 }

